snapcraft/plugins/kbuild.py

```diff
@@ -0,0 +1,181 @@
+import os
+import shutil
+import subprocess
+
+from snapcraft import BasePlugin
+from snapcraft import common
+
+
+logger = logging.getLogger(__name__)
+
+
+class KBuildPlugin(BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+
+        schema['properties']['kdefconfig'] = {
+            'type': 'array',
+            'default': ['defconfig'],
+        }
+
+        schema['properties']['kconfigfile'] = {
+            'type': 'string',
+            'default': None,
+        }
+
+        schema['properties']['kconfigs'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string',
+            },
+            'default': [],
+        }
+
+        schema['properties']['build-parameters'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string',
+            },
+            'default': [],
+        }
+
+        return schema
+
+    def __init__(self, name, options):
+        super().__init__(name, options)
+        self.build_packages.append('make')
+
+        self.make_targets = []
+        self.make_install_targets = ['install']
+        self.make_cmd = [
+            'make', '-j{}'.format(common.get_parallel_build_count())]
+        if self.options.build_parameters:
+            self.make_cmd.extend(self.options.build_parameters)
+        if logger.isEnabledFor(logging.DEBUG):
+            self.make_cmd.append('V=1')
+
+    def do_base_config(self, config_path):
+        # if kconfigfile is provided use that
+        # otherwise use defconfig to seed the base config
+        if self.options.kconfigfile is None:
+            # first run defconfig for setting baseline
+            self.run(self.make_cmd + self.options.kdefconfig)
+        else:
+            shutil.copy(self.options.kconfigfile, config_path)
+
+    def do_patch_config(self, config_path):
+        # prepend the generated file with provided kconfigs
+        #  - concat kconfigs to buffer
+        #  - read current .config and append
+        #  - write out to disk
+        config = '\n'.join(self.options.kconfigs)
+
+        with open(config_path, 'r') as f:
+            config = (
+                '{config_override}\n\n{config}\n{config_override}\n'.format(
+                    config_override=config, config=f.read()))
+
+        # note that prepending and appending the overrides seems
+        # only way to convince all kbuild versions to pick up the
+        # configs during oldconfig in .config
+        config = '{}\n\n{}'.format(config, '\n'.join(self.options.kconfigs))
+
+        with open(config_path, 'w') as f:
+            f.write(config)
+
+    def do_remake_config(self):
+        # update config to include kconfig amendments using oldconfig
+        cmd = 'yes "" | make {} oldconfig'.format(
+            ' '.join(self.options.build_parameters))
```

Comment:
Referring to the [subprocess security guidelines](https://docs.python.org/3/library/subprocess.html#security-considerations), consider using `shlex.quote()` here to escape user-supplied input due to the use of `shell=True` below.